csds-client/client/client.go

```diff
@@ -0,0 +1,255 @@
+	"fmt"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+	"google.golang.org/grpc/metadata"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+type Flag struct {
+	uri         string
+	platform    string
+	authnMode   string
+	apiVersion  string
+	requestFile string
+	requestYaml string
+	jwt         string
+	configFile  string
+	monitorFreq string
+}
+
+type Client struct {
+	cc         *grpc.ClientConn
+	csdsClient csdspb.ClientStatusDiscoveryServiceClient
+
+	nm   []*envoy_type_matcher.NodeMatcher
+	md   metadata.MD
+	info Flag
+}
+
+// ParseFLags parses flags to info
+func ParseFlags() Flag {
+	uriPtr := flag.String("service_uri", "trafficdirector.googleapis.com:443", "the uri of the service to connect to")
+	platformPtr := flag.String("cloud_platform", "gcp", "the cloud platform (e.g. gcp, aws,  ...)")
+	authnModePtr := flag.String("authn_mode", "auto", "the method to use for authentication (e.g. auto, jwt, ...)")
+	apiVersionPtr := flag.String("api_version", "v2", "which xds api major version  to use (e.g. v2, v3 ...)")
+	requestFilePtr := flag.String("request_file", "", "yaml file that defines the csds request")
+	requestYamlPtr := flag.String("request_yaml", "", "yaml string that defines the csds request")
+	jwtPtr := flag.String("jwt_file", "", "path of the -jwt_file")
+	configFilePtr := flag.String("file_to_save_config", "", "the file name to save config")
+	monitorFreqPtr := flag.String("monitor_freq", "", "the frequency of sending request in monitor mode (e.g. 500ms, 2s, 1m ...)")
+
+	flag.Parse()
+
+	f := Flag{
+		uri:         *uriPtr,
+		platform:    *platformPtr,
+		authnMode:   *authnModePtr,
+		apiVersion:  *apiVersionPtr,
+		requestFile: *requestFilePtr,
+		requestYaml: *requestYamlPtr,
+		jwt:         *jwtPtr,
+		configFile:  *configFilePtr,
+		monitorFreq: *monitorFreqPtr,
+	}
+
+	return f
+}
+
+// parseNodeMatcher parses the csds request yaml to nodematcher
+func (c *Client) parseNodeMatcher() error {
+	if c.info.requestFile == "" && c.info.requestYaml == "" {
+		return fmt.Errorf("missing request yaml")
+	}
+
+	var nodematchers []*envoy_type_matcher.NodeMatcher
+	if err := parseYaml(c.info.requestFile, c.info.requestYaml, &nodematchers); err != nil {
+		return err
+	}
+
+	c.nm = nodematchers
+
+	// check if required fields exist in nodematcher
+	switch c.info.platform {
+	case "gcp":
+		switch c.info.uri {
+		case "trafficdirector.googleapis.com:443":
+			keys := []string{"TRAFFICDIRECTOR_GCP_PROJECT_NUMBER", "TRAFFICDIRECTOR_NETWORK_NAME"}
+			for _, key := range keys {
+				if value := getValueByKeyFromNodeMatcher(c.nm, key); value == "" {
+					return fmt.Errorf("Missing field %v in NodeMatcher", key)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// connWithAuth connects to uri with authentication
+func (c *Client) connWithAuth() error {
+	var scope string
+	switch c.info.authnMode {
+	case "jwt":
+		if c.info.jwt == "" {
+			return fmt.Errorf("missing jwt file")
```

Comment:
As there is no formatting required, you can use errors.New() instead. Otherwise using fmt.Errorf suggests formatting was desired, but probably missed?

Reply:
Replaced it by `errors.New()`